contrib/win32/win32compat/tncon.c

```diff
@@ -148,639 +150,676 @@ ReadConsoleForTermEmul(HANDLE hInput, char *destin, int destinlen)
-							NetWriteString2(pParams->Socket, (char *)ALT_PF7_KEY, strlen(ALT_PF7_KEY), 0);
-
-						else if ((dwControlKeyState & SHIFT_PRESSED) && ((dwControlKeyState & RIGHT_ALT_PRESSED) ||
-							(dwControlKeyState & LEFT_ALT_PRESSED)) && ((dwControlKeyState & LEFT_CTRL_PRESSED) ||
-							(dwControlKeyState & RIGHT_CTRL_PRESSED)))
-							NetWriteString2(pParams->Socket, (char *)SHIFT_ALT_CTRL_PF7_KEY, strlen(SHIFT_ALT_CTRL_PF7_KEY), 0);
-
-						else if ((dwControlKeyState & RIGHT_ALT_PRESSED) || (dwControlKeyState & LEFT_ALT_PRESSED) &&
-							((dwControlKeyState & LEFT_CTRL_PRESSED) || (dwControlKeyState & RIGHT_CTRL_PRESSED)))
-							NetWriteString2(pParams->Socket, (char *)ALT_CTRL_PF7_KEY, strlen(ALT_CTRL_PF7_KEY), 0);
-
-						else if ((dwControlKeyState & SHIFT_PRESSED) && ((dwControlKeyState & RIGHT_ALT_PRESSED) ||
-							(dwControlKeyState & LEFT_ALT_PRESSED)))
-							NetWriteString2(pParams->Socket, (char *)SHIFT_ALT_PF7_KEY, strlen(SHIFT_ALT_PF7_KEY), 0);
-
-						else if ((dwControlKeyState & SHIFT_PRESSED) && ((dwControlKeyState & LEFT_CTRL_PRESSED) ||
-							(dwControlKeyState & RIGHT_CTRL_PRESSED)))
-							NetWriteString2(pParams->Socket, (char *)SHIFT_CTRL_PF7_KEY, strlen(SHIFT_CTRL_PF7_KEY), 0);
-						break;
-					case VK_F8:
-						if (dwControlKeyState == 0)
-							NetWriteString2(pParams->Socket, (char *)PF8_KEY, strlen(PF8_KEY), 0);
-
-						else if (dwControlKeyState == SHIFT_PRESSED)
-							NetWriteString2(pParams->Socket, (char *)SHIFT_PF8_KEY, strlen(SHIFT_PF8_KEY), 0);
-
-						else if (dwControlKeyState == LEFT_CTRL_PRESSED || dwControlKeyState == RIGHT_CTRL_PRESSED)
-							NetWriteString2(pParams->Socket, (char *)CTRL_PF8_KEY, strlen(CTRL_PF8_KEY), 0);
-
-						else if (dwControlKeyState == LEFT_ALT_PRESSED || dwControlKeyState == RIGHT_ALT_PRESSED)
-							NetWriteString2(pParams->Socket, (char *)ALT_PF8_KEY, strlen(ALT_PF8_KEY), 0);
-
-						else if ((dwControlKeyState & SHIFT_PRESSED) && ((dwControlKeyState & RIGHT_ALT_PRESSED) ||
-							(dwControlKeyState & LEFT_ALT_PRESSED)) && ((dwControlKeyState & LEFT_CTRL_PRESSED) ||
-							(dwControlKeyState & RIGHT_CTRL_PRESSED)))
-							NetWriteString2(pParams->Socket, (char *)SHIFT_ALT_CTRL_PF8_KEY, strlen(SHIFT_ALT_CTRL_PF8_KEY), 0);
+		ReadConsoleInputW(hInput, inputRecordArray, inputRecordArraySize, &dwInput);
+
+		for (DWORD i=0; i < dwInput; i++) {
+			INPUT_RECORD inputRecord = inputRecordArray[i];
+
+			switch (inputRecord.EventType) {
+			case WINDOW_BUFFER_SIZE_EVENT:
+				queue_terminal_window_change_event();
+				break;
+
+			case FOCUS_EVENT:
+				/* FALLTHROUGH */
+			case MENU_EVENT:
+				break;
+
+			case KEY_EVENT:
+				if (((inputRecord.Event.KeyEvent.bKeyDown) || (!inputRecord.Event.KeyEvent.bKeyDown && inputRecord.Event.KeyEvent.wVirtualKeyCode == VK_MENU)) &&
+				    (inputRecord.Event.KeyEvent.uChar.UnicodeChar != L'\0')) {
+					int n = WideCharToMultiByte(
```

Comment:
It looks like you are attempting the WideCharToMultiByte conversion before you have determined whether it is part of a surrogate pair. This is unnecessary, and can be moved to line 215.